datafusion/substrait/src/consumer.rs

```diff
@@ -679,12 +680,46 @@ pub async fn from_substrait_rex(
                 }
             }
         }
+        Some(RexType::Cast(cast)) => Ok(Arc::new(Expr::Cast(Cast::new(
+            Box::new(
+                from_substrait_rex(
+                    cast.as_ref().input.as_ref().unwrap().as_ref(),
+                    input_schema,
+                    extensions,
+                )
+                .await?
+                .as_ref()
+                .clone(),
+            ),
+            from_substrait_type(cast.as_ref().r#type.as_ref().unwrap())?,
```

Comment:
It would be good to add comments about the `unwrap` calls, if they are guaranteed always to work. I can't tell just from looking at the diff here if they are safe or not.

Reply:
Good point. Generally if the Substrait plan has been produced correctly, the `cast` expression should always have an output type. But since we cannot guarantee that, I added a match arm to produce appropriate error in [58f90ce](https://github.com/apache/arrow-datafusion/pull/5137/commits/58f90ce085583ba392ed6bbc4db90e7b0b83656b),